src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,167 +23,344 @@ using namespace opossum::expression_functional;  // NOLINT
+  }
+}
+
+void try_join_to_semi_rewrite(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    const std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node) {
+  // Sometimes, joins are not actually used to combine tables but only to check the existence of a tuple in a second
+  // table. Example: SELECT c_name FROM customer, nation WHERE c_nationkey = n_nationkey AND n_name = 'GERMANY'
+  // If the join is on a unique/primary key column, we can rewrite these joins into semi joins. If, however, the
+  // uniqueness is not guaranteed, we cannot perform the rewrite as non-unique joins could possibly emit a matching
+  // line more than once.
 
-    if (const auto stored_table_node = std::dynamic_pointer_cast<StoredTableNode>(node)) {
-      stored_table_node->set_pruned_column_ids(pruned_column_ids);
-    } else if (const auto mock_node = std::dynamic_pointer_cast<MockNode>(node)) {
-      mock_node->set_pruned_column_ids(pruned_column_ids);
-    } else {
-      // We don't know how to prune columns from this leaf-node type (CreateViewNode, etc.), so do nothing
+  auto join_node = std::dynamic_pointer_cast<JoinNode>(node);
+  if (join_node->join_mode != JoinMode::Inner) return;
+
+  // Check whether the left/right inputs are actually needed by following operators
+  auto left_input_is_used = false;
+  auto right_input_is_used = false;
+  for (const auto& output : node->outputs()) {
+    for (const auto& required_expression : required_expressions_by_node.at(output)) {
+      if (expression_evaluable_on_lqp(required_expression, *node->left_input())) left_input_is_used = true;
+      if (expression_evaluable_on_lqp(required_expression, *node->right_input())) right_input_is_used = true;
     }
+  }
+  DebugAssert(left_input_is_used || right_input_is_used, "Did not expect a useless join");
+  if (left_input_is_used && right_input_is_used) return;
+
+  // Check whether the join predicates operate on unique columns.
+  auto join_is_unique_on_left_side = false;
+  auto join_is_unique_on_right_side = false;
+
+  const auto& join_predicates = join_node->join_predicates();
+  for (const auto& join_predicate : join_predicates) {
+    const auto is_unique_column = [](const auto& expression) {
+      const auto& column = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+      if (!column) return false;
 
-    return LQPVisitation::VisitInputs;
-  });
+      const auto& column_reference = column->column_reference;
+      const auto& stored_table_node =
+          std::dynamic_pointer_cast<const StoredTableNode>(column_reference.original_node());
+      if (!stored_table_node) return false;
+
+      const auto& table = Hyrise::get().storage_manager.get_table(stored_table_node->table_name);
+      for (const auto& table_constraint : table->get_soft_unique_constraints()) {
+        // This currently does not handle multi-column constraints, but that should be easy to add once needed.
```

Comment:
The emphasis is on the second part. Otherwise people might think that there is some god-given limitation that prevents multi-column constraints from being useful here.